src/librvnpal/src/posix/virtualmemory.c

```diff
@@ -0,0 +1,68 @@
+#if defined(__unix__) || defined(__APPLE__)
+
+#define _GNU_SOURCE
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <assert.h>
+
+#include "rvn.h"
+#include "status_codes.h"
+
+int32_t
+rvn_prefetch_virtual_memory(void *virtualAddress, int64_t length, int32_t *detailed_error_code)
+{
+    int32_t rc = madvise(virtualAddress, length, MADV_WILLNEED);
+    if (rc != SUCCESS)
+        *detailed_error_code = errno;
+    return rc;
+}
+
+int32_t
+rvn_prefetch_ranges(struct RVN_RANGE_LIST *range_list, int32_t count, int32_t *detailed_error_code)
+{
+    int32_t i = 0;
+    int32_t rc = 0;
+    int32_t prefetch_errno = 0;
+    for (i = 0; i < count; i++)
+    {
+        struct RVN_RANGE_LIST record = range_list[i];
+        int32_t prefetch_rc = rvn_prefetch_virtual_memory(record.virtual_address, record.number_of_bytes, &prefetch_errno);
```

Comment:
```suggestion
        int32_t prefetch_rc = rvn_prefetch_virtual_memory(record.virtual_address, record.number_of_bytes, detailed_error_code);
```

Reply:
I want to record the first error but continue with the other ranges. The above suggestion will override the errno if second iteration successful.

Reply:
No, it is a bad idea. We want to exit early

Reply:
We got an error, and continuing work from there means that we don't know what is our state.
given that this is an optimization only, that doesn't matter